src/backend/distributed/commands/create_distributed_table.c

```diff
@@ -1665,6 +1808,51 @@ UndistributeTable(Oid relationId)
 							NULL, None_Receiver, NULL);
 	}
 
+	if (conversionType == ALTER_DISTRIBUTED_TABLE || (conversionType == ALTER_TABLE_SET_ACCESS_METHOD && IsCitusTableType(relationId, DISTRIBUTED_TABLE)))
+	{
+		Var *distributionKey = NULL;
+
+		if (distributionColumn)
+		{
+			Relation relation = try_relation_open(relationId, ExclusiveLock);
+			relation_close(relation, NoLock);
+			distributionKey = BuildDistributionKeyFromColumnName(relation,
+																 distributionColumn);
+		}
+		else
+		{
+			distributionKey = DistPartitionKey(relationId);
+		}
+
+		if (colocateWith == NULL)
+		{
+			Var *originalDistributionKey = DistPartitionKey(relationId);
+			if ((distributionColumn == NULL || originalDistributionKey->vartype == distributionKey->vartype) && shardCountIsNull)
+			{
+				StringInfo qualifiedRelationName = makeStringInfo();
+				appendStringInfo(qualifiedRelationName, "%s.%s", schemaName, relationName);
+				colocateWith = qualifiedRelationName->data;
+			}
+			else
+			{
+				colocateWith = "default";
+			}
+		}
+		char partitionMethod = PartitionMethod(relationId);
+		CreateDistributedTable(get_relname_relid(tempName, schemaId), distributionKey, partitionMethod, shardCount, colocateWith, false);
+	}
+	else if (conversionType == ALTER_TABLE_SET_ACCESS_METHOD)
+	{
+		if (IsCitusTableType(relationId, REFERENCE_TABLE))
+		{
+			CreateDistributedTable(get_relname_relid(tempName, schemaId), NULL, DISTRIBUTE_BY_NONE, ShardCount, NULL, false);
+		}
+		else if (IsCitusTableType(relationId, CITUS_LOCAL_TABLE))
+		{
+			CreateCitusLocalTable(get_relname_relid(tempName, schemaId));
+		}
+	}
```

Comment:
it might be nice to move this block into its own function, something like:

```c
if (conversionType == ALTER_DISTRIBUTED_TABLE)
{
    CreateDistributedTableLike(newRelationId, oldRelationId, distributionColumn);
}
else if (conversionType == ALTER_TABLE_SET_ACCESS_METHOD)
{
    CreateCitusTableLike(newRelationId, oldRelationId);
}


/*
 * CreateCitusTableLike converts table with the given name to a Citus table
 * in a way that matches the likeRelationId.
 */
static void
CreateCitusTableLike(Oid relationId, Oid likeRelationId, char *newDistributionColumn, int newShardCount, char *newColocateWith)
{
	if (IsCitusTableType(likeRelationId, DISTRIBUTED_TABLE))
	{
		CreateDistributedTableLike(relationId, likeRelationId, newDistributionColumn, newShardCount, newColocateWith);
	}
    else if (IsCitusTableType(likeRelationId, REFERENCE_TABLE))
	{
		CreateDistributedTable(relationId, NULL, DISTRIBUTE_BY_NONE, ShardCount, NULL, false);
	}
	else if (IsCitusTableType(likeRelationId, CITUS_LOCAL_TABLE))
	{
		CreateCitusLocalTable(relationId);
	}
}
```

Reply:
CreateDistributedTableLike left as an exercise to the reader :)  but pretty much corresponding to the first if block